src/appleseed.shaders/include/appleseed/maya/as_maya_fractal_helpers.h

```diff
@@ -0,0 +1,380 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+ 
+#ifndef AS_MAYA_FRACTAL_HELPERS_H
+#define AS_MAYA_FRACTAL_HELPERS_H
+
+#define MAYA_LATTICE_SIZE   20
+
+#include "appleseed/fractal/as_noise_helpers.h"
+#include "appleseed/math/as_math_helpers.h"
+
+void implode_2d(
+    float implode,
+    float implode_center[2],
+    output float x,
+    output float y)
+{
+    if (implode > EPS || implode < -EPS)
+    {
+        x -= implode_center[0];
+        y -= implode_center[1];
+
+        float dist = hypot(x, y);
+
+        if (dist > EPS)
+        {
+            float factor = pow(dist, 1 - implode) / dist;
+            x *= factor;
+            y *= factor;
+        }
+        x += implode_center[0];
+        y += implode_center[1];
+    }
+}
+
+float maya_turbulence(
+    point surface_point,
+    float initial_time,
+    float filter_width,
+    float amplitude,
+    int octaves,
+    float lacunarity,
+    float gain)
+{
+    point xyz = surface_point;
+    float amp = amplitude, filter_size = filter_width;
+    float current_time = initial_time, sum = 0.0;
+
+    for (int i = 0; i < octaves; ++i)
+    {
+        if (!amp)
+        {
+            break;
+        }
+
+        // The base frequency looks too regular, perturb it with point+noise.
+        float tmp = amp *
+            filtered_snoise(
+                xyz + noise("uperlin", lacunarity),
+                current_time,
+                filter_size);
+
+        sum += abs(tmp);
+        amp *= gain;
+        xyz *= lacunarity;
+
+        filter_size *= lacunarity;
+        current_time *= lacunarity;
+    }
+    return sum;
+}
+
+float maya_fBm(
+    point surface_point,
+    float initial_time,
+    float filter_width,
+    float amplitude,
+    int octaves,
+    float lacunarity,
+    float gain)
+{
+    point xyz = surface_point;
+    float amp = amplitude, filter_size = filter_width;
+    float current_time = initial_time, sum = 0.0;
+
+    for (int i = 0; i < octaves; ++i)
+    {
+        if (!amp)
+        {
+            break;
+        }
+
+        // These factors seem to match Maya's result better.
+        sum += amp * 1.2 *
+            (filtered_snoise(xyz, current_time, filter_size) - 0.1) + 0.05;
+
+        amp *= gain;
+        xyz *= lacunarity;
+
+        filter_size *= lacunarity;
+        current_time *= lacunarity;
+    }
+    return clamp(sum * 0.5 + 0.5, 0.0, 1.0);
+}
+
+float maya_cos_waves_2d(
+    point surface_point,
+    float current_time,
+    int current_step,
+    int waves)
+{
+    int seed = current_step * 50;
+
+    float x, y, h, out = 0.0;
+
+    for (int i = 0; i < waves; ++i)
+    {
+        do
+        {
+            x = random_noise(seed++);
+            y = random_noise(seed++);
+            h = hypot(x, y);
+        }
+        while (h <= 0.0);
```

Comment:
This looks strange: `h` is almost never going to be <= 0.

Rejection sampling (since this is what this loop implements) is usually rejecting samples outside the unit disk in order to implement uniform disk sampling.

Just checking that this was intentional and not a buggy disk sampling algorithm :)